### DISCUSSION TITLE
reading-json-and-binary-data-from-multipart-form-data-sections-in-aspnetcore/

### DISCUSSION THREAD

**Discussion by giscus** [andrewlock/blog-comments#395] [Announcements]
# reading-json-and-binary-data-from-multipart-form-data-sections-in-aspnetcore/

In this post I describe how to read both JSON and binary data from a multipart/form-data request using two approaches - MultipartReader and IFormFile

https://andrewlock.net/reading-json-and-binary-data-from-multipart-form-data-sections-in-aspnetcore/

**Comment by moggoly** [comment]
I'm reading this, and struggling to understand why it's not working for you - as I've got this exact scenario working perfectly well in both a Razor app and in an API. Could you tell me what I'm doing differently in the below code from what your issue is? The only difference I can see is that I'm binding to a ```List<IFormFile>``` instead of a single ```IFormFile```.

```csharp
[BindProperty]
public AddPostInputModel AddPostInput { get; set; }

public class AddPostInputModel
{
    public string Title { get; set; }
    public string Description { get; set; }
    public List<IFormFile> Images { get; set; }
}
```

```html
<form class="new-post-form" method="post" enctype="multipart/form-data">
	<label for="title-input">Description</label>
	<input id="title-input" type="text" asp-for="AddPostInput.Title" />
	<label for="description-input">Description</label>
	<input id="description-input" type="text" asp-for="AddPostInput.Description" />
	<label for="image-input">Images</label>
	<input id="image-input"
    	type="file"
    	accept="image/*"
    	asp-for="AddPostInput.Images" />
	<button class="btn btn-primary" type="submit">Submit</button>
</form>
```

**Reply by andrewlock** [reply]
Sure, this is a fundamentally different scenario😃 

The data I'm reading is _not_ sent from an HTML form. It's sent from an `HttpClient`. That means the structure of the data is different. Your requests will look something like this:

```http
------WebKitFormBoundary8Vq2TDi66aYi2H5d
Content-Disposition: form-data; name="Title"

The title
------WebKitFormBoundary8Vq2TDi66aYi2H5d
Content-Disposition: form-data; name="Description"

The description
------WebKitFormBoundary8Vq2TDi66aYi2H5d
Content-Disposition: form-data; name="myFile"; filename="some-file-to-send.txt"
Content-Type: application/octet-stream

<binary data>

------WebKitFormBoundary8Vq2TDi66aYi2H5d--
```

Whereas the data I was receiving was like this:

```http
--73dc24e0-b350-48f8-931e-eab338df00e1
Content-Type: application/json; charset=utf-8
Content-Disposition: form-data; name=myJson

{"Name":"Andrew","Type":"Engineer"}
--73dc24e0-b350-48f8-931e-eab338df00e1
Content-Type: application/octet-stream
Content-Disposition: form-data; name=binary

<binary data>
--73dc24e0-b350-48f8-931e-eab338df00e1--
```

i.e. 
- No `filename` in `Content-Disposition`
- JSON data `Content-Type`, not the default `text/plain`

**Reply by moggoly** [reply]
Gotchya. I initially read this article as being an issue with binding (which it is fundamentally), but really it's more about how to get HttpClient working properly with the built-in binding, that's what confused me. Thanks for the clarification.

**Reply by andrewlock** [reply]
Yeah exactly, it's really a question of how to bind to a specific (non-standard) request body! 🙂

**Comment by thomaslevesque** [comment]
I solved a similar problem a few years ago, using a custom model binder. Not sure if it could be adapted to your use case...
https://thomaslevesque.com/2018/09/04/handling-multipart-requests-with-json-and-file-uploads-in-asp-net-core/

**Comment by binraider** [comment]
Andrew, i just read this article :  https://brokul.dev/sending-files-and-additional-data-using-httpclient-in-net-core, which uses custom binding - do you know if this technique would work? Is it an approach you have looked at?
Regards,
Marshall

**Comment by unsafePtr** [comment]
I've run into interesting problem. There is ContentType multipart section header. `IFormFile` has property with the same name, and our implementation is relying on this. However, for some reason when uploading `.xlsm` file ContentType is `application/octet-stream`, while when uploading `.xlsx` file ContentType is set correctly.

**Comment by JarrodJ83** [comment]
Have you ever had to _**return**_ a response that contained binary and json? I'm facing this now and not a lot of info out there on the topic.
